content/blog/2020/how-selenium-works-2-navigation.md

```diff
@@ -0,0 +1,70 @@
+{{< /highlight >}}
+
+We can see above... it looks easy enough right... no!
+
+Actually, it leads me to a pet peeve in interviews. If someone ever asks you to ever describe what happens in a browser when you type a URL and press enter there is a high probability that they have no real idea what goes into navigation. Anyway... back to Selenium and how we navigate.
+
+# `driver.get`
+
+Once we send the request over the [transportation layer]({{< relref "/blog/how-selenium-works-transport">}}) the real fun starts. We need to figure out where we want to go and tell the browser to start going there for us. Unfortunately this will lead to the first of the many problems for automators.
+
+## Certificates
+
+If you have ever worked in major corporate companies where technology is not the core then you will understand the pain that a development team might have to endure. Now double that pain and you're starting to get an idea of the pain that testers and automators have to deal with.
+
+One of these pains is certificates. Companies are cheap and will do self-signed certificates and other monstrosities. Especially in the early days of Selenium when there wasn't services like [Let's Encrypt](https://letsencrypt.org/). And even now, most Developers and QA teams rarely have access to change configurations on their test environments or on their CI servers. We needed to come up with a way to circumvent the certificates. (_This is one of the first reasons why Selenium is seen as a security risk_)
+
+So each of the browsers will allow the poorly configured certificates through when we automate. It's because if we didn't a lot of testers/developers would never be able to test their sites. 
+
+Now... that we go round the first problem we need to move on to getting the page to load.
+
+## Loading
+
+Once we have got round the certificates we get the page to load. Luckily we don't need to do anything more than the equivalent of 
+
+{{< highlight javascript >}}
+location = "https://www.theautomatedtester.co.uk"; 
+// or
+window.location.href = "https://www.theautomatedtester.co.uk";
+{{< /highlight >}}
+
+## Done... 
+
+When Selenium has "finished" a command it will return. So, we just need to wait for the page to be finished loading. To be honest, what does _"finished loading"_ even mean.
+
+In the browser it will fire a number of different events. We will know when the page has been shown and then tell you what it's [`readyState`](https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState). Selenium will check all of that and it will also wait for the [`DOMContentLoaded`](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event).
+
+Once we have done those checks, and you can manipulate how we look at those events, if you want it to load quicker, using the [Page Load Strategies](https://w3c.github.io/webdriver/#dfn-page-loading-strategy). This is more of a power user feature so I wouldn't worry about them now.
```

Comment:
It might be interesting to delve into the motivation behind the page loading strategies (namely that they allow people to have a lot more control about when commands execute, at the cost of making them more responsible for making sure that the app being controlled is ready for the next step)